### PR TITLE
Tuning the CIFAR-10 classifier with optuna, ray-tune and ASHAScheduler.

### PR DESCRIPTION
**Pull request by SagnikDe2024**
The focus will be to find those hyper parameters which reduces the number of tunable parameters and simultaneously decreases validation loss.

### CODE REVIEW

File: src/models/codec.py [hunk 1]
```diff
@@ -1,50 +1,19 @@
...
+def generate_separated_kernels(input_channel: int, output_channel: int, k_size: int, a: float = (1 / 2), r: float = 0.0,
 							   add_padding: bool = True, bias=False, stride: int = 1):
 	c_in = input_channel
 	c_out = output_channel
 	t = c_out / c_in
 	k = k_size
 
 	if 0 < r < 1 and t != 1:
-		a = log2(c_in * k ** 2 * r + r - 1) / log2(t) - log2(c_in * k * (t + 1) + 1) / log2(t) + 1
+		a = log2((t + 1) / (t * r * k)) / log2(1 / t)
```
Comment (SagnikDe2024) [inline src/models/codec.py:16]: CNNs will have the bias removed and instead batch norm bias will be used from now on.

File: src/models/codec.py [hunk 2]
```diff
@@ -53,55 +22,125 @@ def generate_separated_kernels(k_size: int, input_channel: int, output_channel:
...
 
-	conv_layer_1 = nn.Conv2d(c_in, c_intermediate, (k, 1), padding=(padding, 0), bias=bias,stride=stride)
-	conv_layer_2 = nn.Conv2d(c_intermediate, c_out, (1, k), padding=(0, padding), bias=bias,stride=stride)
+	conv_layer_1 = nn.Conv2d(c_in, c_intermediate, (1, k), padding=(0, padding), bias=bias, stride=(1, stride))
+	conv_layer_2 = nn.Conv2d(c_intermediate, c_out, (k, 1), padding=(padding, 0), bias=bias, stride=(stride, 1))
+
 	return conv_layer_1, conv_layer_2
 
 
-def conv_kernel(k_size: int, input_channel: int, output_channel: int, separable, ratio):
```
Comment (SagnikDe2024) [inline src/models/codec.py:61]: Not used anymore.

File: src/models/codec.py [hunk 3]
```diff
@@ -111,66 +150,5 @@ def __init__(self, input_size, output_size, kernel_sizes=None, channels=None, la
...
-			channels = [*inp_out_channels]
-		print(f'Decoder channels and kernels: {channels},{kernel_sizes}')
-		dec = Decoder(input_size, output_size, kernel_sizes, channels)
-		return dec
-
 	def forward(self, latent_z):
 		return self.sequence.forward(latent_z)
-
-
-class Encoder(nn.Module):
```
Comment (SagnikDe2024) [inline src/models/codec.py:133]: Actual class is in classifier now.

File: src/models/codec.py [hunk 4]
```diff
@@ -111,66 +150,5 @@ def __init__(self, input_size, output_size, kernel_sizes=None, channels=None, la
 				sequence.append(activation_layer)
 		self.sequence = nn.Sequential(*sequence)
 
-	@classmethod
```
Comment (SagnikDe2024) [inline src/models/codec.py:114]: Nicer decode will be used from now on.

File: src/tune_classifier.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+from ray.tune.search.optuna import OptunaSearch
+from torchvision.datasets import CIFAR10
+from torchvision.transforms import transforms
+
+from src.classifier.classifier import Classifier
+from src.utils.common_utils import AppLog
+from src.wip.training import ExperimentModels
+
+
+@torch.compiler.disable(recursive=True)
```
Comment (SagnikDe2024) [inline src/tune_classifier.py:20]: Issues with torch compile ?

File: src/tune_classifier.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+		hashed = f'{hash(param_s)}'
+		self.dir_num += 1
+		return f'3_{save_time}_{hashed}_{self.dir_num}'
+
+	def tune_classifier_model(self, restore=True):
+
+		tempdir = self.working_dir / 'out'
+		ray.init(_temp_dir=f'{tempdir.absolute()}')
+
+		if restore and Tuner.can_restore(self.checkpoint_store):
```
Comment (SagnikDe2024) [inline src/tune_classifier.py:84]: Does not restore with optuna for for some reason.

File: src/visualize_classifier_results.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+		if len(stored_checkpoint_dirs) == 0:
+			print(f'No checkpoint directory found for {sub_directory}')
+			stored_checkpoint_files = [chk for chk in sub_directory.iterdir() if chk.is_file()]
+			for file in stored_checkpoint_files:
+				file.unlink()
+			sub_directory.rmdir()
+			continue
+
+
+def find_classify_checkpoint():
```
Comment (SagnikDe2024) [inline src/visualize_classifier_results.py:40]: Need more comments. Will merge for now.

**Review by SagnikDe2024 — COMMENTED** [review]
A big merge but the classifier is working as needed.
